### PR TITLE
Fix browser / version detection for firefox

### PR DESCRIPTION
**Pull request by matrikular**
### Summary of Changes
Change the regular expression to detect the correct Firefox version.

### Testing Instructions
Paste the following code snippet into the [index.php of the protostar template](https://github.com/joomla/joomla-cms/blob/staging/templates/protostar/index.php#L13) and open the page in Firefox:
```
jimport('joomla.environment.browser');

$navigator = JBrowser::getInstance();

var_dump($navigator->getAgentString(), $navigator->getVersion());
```
### Expected result
The result should show the correct Firefox version e.g.
```
array (size=2)
  'UserAgent:' => string 'Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:54.0) Gecko/20100101 Firefox/54.0' (length=78)
  'Version:' => string '54.0' (length=4)
```
### Actual result
The version number is "always" 5.0.
```
array (size=2)
  'UserAgent:' => string 'Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:54.0) Gecko/20100101 Firefox/54.0' (length=78)
  'Version:' => string '5.0' (length=3)
```

Edit: maybe the agent variable should be changed from "mozilla" to "firefox", too? Can we do this for 3.8 or not  until 4.x?

### CODE REVIEW

**Comment by ghost** [comment]
I have tested this item :white_check_mark: successfully on 886dd2524661cd44a4c32040086f30ee74b38fe4
Without PR:
without
With PR:
with

<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17051">issues.joomla.org/tracker/joomla-cms/17051</a>.</sub>

**Comment by gorgonz** [comment]
verified successfully with linux, result is 
string(68) "Mozilla/5.0 (X11; Linux x86_64; rv:52.0) Gecko/20100101 Firefox/52.0" string(3) "5.0"

**Comment by ghost** [comment]
@gorgonz i have altered your Test at Issue Tracker.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17051">issues.joomla.org/tracker/joomla-cms/17051</a>.</sub>

**Comment by ghost** [comment]
RTC after two successful tests.

**Comment by matrikular** [comment]
Before this gets merged I would love to get some feedback as to whether or not to change the agent variable to firefox instead of mozilla.

@mbabker @wilsonge thoughts?

File: libraries/joomla/environment/browser.php [hunk 1]
```diff
@@ -243,7 +243,7 @@ public function match($userAgent = null, $accept = null)
 
 			// We have to check for Edge as the first browser, because Edge has something like:
 			// Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/51.0.2704.79 Safari/537.36 Edge/14.14393
-			if (preg_match('|Edge/([0-9.]+)|', $this->agent, $version))
+			if (preg_match('|Edge\/([0-9.]+)|', $this->agent, $version))
```
Comment (Quy) [inline libraries/joomla/environment/browser.php]: Do not need to escape `/` unless the first/last `|` is `/`.

Comment (matrikular) [reply libraries/joomla/environment/browser.php]: @Quy To make it consistent and safe to test anywhere else, I thought it might be a good idea and causes no harm.

**Comment by wilsonge** [comment]
Are there any user agents with Mozzila in that don't have FF in

**Comment by mbabker** [comment]
@matrikular any feedback on George's question?

**Comment by matrikular** [comment]
@mbabker As mentioned in the description / comments both here and in your PR, "all" User Agents have got Mozilla in it. For that, our browser class would currently identify every browser as "Mozilla/5.0" (mozilla) if no other (more strict) rule applies before, of course.

If the browser is a clone / derivative of Firefox, you could find something like the following in the User Agent string: "Mozilla/5.0 (Windows NT 5.2; Win64; x64; rv:39.0) Gecko/20100101 Firefox/39.0 Waterfox/39.0".
In this example, adding a regular expression for Firefox alone would match and return "firefox", even though it is Waterfox or another clone / derivative (e.g Pale Moon).

In short; Yes, there are User Agents with Mozilla in that don't have FF in. Sadly, if we want to get it right, it seems that [this is not easily covered](https://github.com/sinergi/php-browser-detector/blob/master/src/BrowserDetector.php#L16) with adding yet another regular expression / rule. Well, at least not for the long run.

That said, I would like to propose the following changes:
- Move the regular expression for Mozilla to the very bottom / end
- Add regular expressions for Firefox and some of the known clones / derivatives
- Check and re-order the existing regular expressions to make sure that we don't have any "false positives"
- [...]

**Comment by mbabker** [comment]
Those changes sound reasonable to me.

**Comment by wilsonge** [comment]
Me too :)

**Comment by mbabker** [comment]
To merge this PR, let's do 3 things:

1) Re-sync with staging
2) Re-add the `elseif (preg_match('|Mozilla/([0-9.]+)|', $this->agent, $version))` condition as the last condition in the browser check
3) In `JBrowserTest` remove [lines 154 thru 157](https://github.com/joomla/joomla-cms/blob/staging/tests/unit/suites/libraries/joomla/environment/JBrowserTest.php#L154-L157) as with this PR the Firefox UAs I added for the tests there should now correctly be parsed

We can work on further improving from here but we really REALLY should (finally) fix Firefox detection.

**Review by wilsonge — APPROVED** [review]
